### DISCUSSION TITLE
OSError: (-1, 'Connection closed by host') on ESP8266 using mqtt_as

### DISCUSSION THREAD

**Discussion by wolfelias** [micropython/micropython#13142] [ESP8266]
Hi! I am using @peterhinch's "mqtt_as" library and deployed it using cross-compile. I ran different examples on an ESP32 which all work fine. Running the same scripts (including the ESP8266 SSL/TLS example in the library) on an ESP8266 results in the folling Error: 

`OSError: (-1, 'Connection closed by host')`

This is the call stack: 

```
>>> 
MPY: soft reboot
#10 ets_task(4020f544, 28, 3fff9ec0, 10)
Checking WiFi integrity.
Got reliable connection
Connecting to broker.
Traceback (most recent call last):
  File "boot.py", line 50, in <module>
  File "uasyncio/core.py", line 1, in run
  File "uasyncio/core.py", line 1, in run_until_complete
  File "uasyncio/core.py", line 1, in run_until_complete
  File "boot.py", line 24, in main
  File "mqtt_as.py", line 645, in connect
  File "mqtt_as.py", line 314, in _connect
  File "mqtt_as.py", line 221, in _as_read
OSError: (-1, 'Connection closed by host')
MicroPython v1.20.0 on 2023-04-26; ESP module with ESP8266
```

Any ideas what this could relate to?

**Comment by zcattacz** [comment]
If your devices are online at the same time, connecting with identical credentials (client_id, user_name etc), make sure the server allows clients to do so. It may be the server is kicking the client off. 
Unless you can read server log, this is not obvious, until figured out.

**Reply by wolfelias** [reply]
Hey thanks for you answer! In my setup the devices connected with different client_ids. I tested the same setup using multiple ESP32 with mqtt_as and also tested the setup using the standart micropython mqtt client, which all worked. Thus I suppose the problem is not the server but the implementation of mqtt_as!

**Reply by zcattacz** [reply]
Hi, I have a stripped down version of mqtt_as which is [runnable on cpython](https://github.com/zcattacz/mqtt_as). It's easier to debug with wireshark. The downside/advantage depending on your need, is that all network management code has been removed, which why it's runable on cpython :-).

**Reply by wolfelias** [reply]
Hey, that looks very good, thank you! Have you had the chance to test it on esp8266 as well?

**Reply by zcattacz** [reply]
Hi, nope, I didn't test it. it might work. But when possible further development is factored in, it may soon become a huge challenge. It's used here on multiple linux boxes and esp32/s2 boards, both have plenty memory. To accommodate esp8266, many changes become impossible. I use arduino pubsub on esp8266 for some project, it worked pretty stable, but micropython is easier to prototype.
